File: daily_problem/2023-3/1625.py
"""
1625. 执行操作后字典序最小的字符串
https://leetcode.cn/problems/lexicographically-smallest-string-after-applying-operations/description/
"""


class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        q = deque([s])
        state = {s}
        res = s
        while q:
            cur = q.popleft()
            res = min(res, cur)
            t1 = ''.join(str((int(x) + a) % 10) if i &
                         1 else x for i, x in enumerate(cur))
            t2 = cur[-b:] + cur[:-b]
            for t in (t1, t2):
                if t not in state:
                    state.add(t)
                    q.append(t)
        return res

File: daily_problem/2023-3/1625.cpp
/*
1625. 执行操作后字典序最小的字符串
https://leetcode.cn/problems/lexicographically-smallest-string-after-applying-operations/description/
*/

#include <algorithm>
#include <deque>
#include <string>
#include <unordered_set>

using std::deque;
using std::min;
using std::string;
using std::unordered_set;

class Solution {
public:
  string findLexSmallestString(string s, int a, int b) {
    deque<string> q{{s}};
    unordered_set<string> state{{s}};
    string res = s;
    int n = s.size();
    while (!q.empty()) {
      string cur = q.front();
      q.pop_front();
      res = min(res, cur);
      string t1 = cur;
      for (int i = 1; i < n; i += 2) {
        t1[i] = (cur[i] - '0' + a) % 10 + '0';
      }
      string t2 = cur.substr(n - b, b) + cur.substr(0, n - b);
      for (const string &t : {t1, t2}) {
        if (!state.count(t)) {
          state.insert(t);
          q.emplace_back(t);
        }
      }
    }
    return res;
  }
};
